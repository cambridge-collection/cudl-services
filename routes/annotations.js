'use strict';

var fs = require('fs');

var express = require('express');
var js2xmlparser = require('js2xmlparser');
var _ = require('lodash');
var csvstringify = require('csv-stringify');
var Q = require('q');
var accepts = require('accepts');

var db = require('../lib/db');
var items = require('../lib/items');
var tagsRoute = require('./tags');

var router = express.Router();

router.get('/:classmark.:ext(json|xml|txt|csv)', function(req, res) {
    sendAnnotationResponse(req, res, {type: req.params.ext});
});

router.get('/:classmark', function(req, res) {
    sendAnnotationResponse(req, res);
});

function extractOptions(req) {
    return {
        type: req.params.ext
    };
}

function sendAnnotationResponse(req, res, options) {
    options = _.defaults({}, options, extractOptions(req));

    Q.all([
        // Returns list of tags
        getAnnotations(req.params.classmark),
        // Does not return value, only throws if tag doesn't exist
        items.ensureItemExists(req.params.classmark)
    ])
    .spread(getNegotiatedResponse(req, res, options.type))
    .then(tagsRoute.sendResponse(res))
    .catch(tagsRoute.handleErrors(res))
    .done();
}

var ANNOTATIONS_BY_ITEM = fs.readFileSync(
    require.resolve('../sql/annotations-by-item.sql'), 'utf-8');

/**
 * Get a promise resolving to the list of annotations for the item with id.
 */
function getAnnotations(id) {
    return db.query(ANNOTATIONS_BY_ITEM, [id])
    .then(function(result) {
        return {
            id: id,
            annotations: result.rows
        };
    });
}


function getNegotiatedResponse(req, res, fixedType) {
    return function(result) {
        var accept = accepts(req);

        var negotiatedType = fixedType || accept.type(['json', 'xml']);

        switch(negotiatedType) {
            case 'json':
                return {
                    type: 'application/json',
                    body: JSON.stringify({
                        id: result.id,
                        annotations: result.annotations,
                        count: result.annotations.length
                    })
                };
            case 'xml':
                var xml = {
                    '@': {
                        count: result.annotations.length,
                        id: result.id
                    },
                    annotation: _.map(result.annotations, function(annotation) {
                        return {
                            '@': {
                                'page': annotation.page
                            },
                            '#': annotation.value,
                            '=': annotation.type
                        };
                    })
                };

                return {
                    type: 'application/xml',
                    body: js2xmlparser('annotations', xml)
                };
            case 'csv':
                /* falls through */
            case 'txt':
                /* falls through */
            default:
                var rows = _.map(result.annotations, function(annotation) {
                    return [annotation.type, annotation.value, annotation.page];
                });

                return Q.nfcall(csvstringify, rows, {
                    header: true,
                    columns: ['type', 'value', 'page']
                })
                .then(function(csv) {
                    return {
                        type: negotiatedType == 'csv' ? 'text/csv' : 'text/plain',
                        body: csv
                    };
                });
        }
    };
}

module.exports = {
    router: router
};
