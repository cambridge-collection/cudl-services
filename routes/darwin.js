var express = require('express');
var request = require('request');

var router = express.Router();

module.exports = function(passport) {

	router.get('/*', passport.authenticate('token', { session: false }),
		function(req, res) {
			var url = config.darwinXTF + req.url + ";idx-ignore=dcpPublic";
			console.log(url);
			req.pipe(request(url)).pipe(res);
		}
	);

	return router;
};
