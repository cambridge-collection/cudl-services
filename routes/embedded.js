var express = require('express');

var router = express.Router();

/* GET home page. */
router.get('/', function(req, res) {
  res.render('index', { title: 'Metadata' });
});

router.get('/:id', function(req, res) {
	var images = [];
	images.push("http://found-dev-pres01.lib.cam.ac.uk/content/images/MS-DAR-00100-000-00001.dzi")
	console.log('laa');
	res.render('player', { title: req.params.id,
			       images: JSON.stringify(images)
	});
	
});

module.exports = router;
