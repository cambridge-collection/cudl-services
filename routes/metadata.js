var express = require('express');
var fs = require("fs");
var util = require("util");

var config = require('../config/base');
var serviceUtil = require("../util.js");

var CORS_HEADERS = {
	"Access-Control-Allow-Origin": "*"
};

var router = express.Router();

/* GET home page. */
router.get('/', function(req, res) {
    res.status(401).send('Unathorised');
/*  res.render('index', { title: 'Unathorised' });*/
});

router.get('/:format/:id',
	function(req, res) {
		// We always want to allow remote ajax access
		res.set(CORS_HEADERS);

		// The response depends on the Origin header, as we block access to
		// non-embeddable items from non-cudl origins. If we don't set
		// Vary: Origin then a response for a client on CUDL could be used by a
		// cache to service a request from an external site.
		res.set("Vary", "Origin");

		// Ensure our ID doesn't contain slashes, as it's used to build
		// filesystem paths.
		if(/\//.test(req.params.id)) {
			res.status(400).json({
				error: util.format("Bad id: %s", req.params.id)
			});
			return;
		}

		var path;

		if (req.params.format == 'json') {
			path = config.dataDir + '/json/' + req.params.id + '.json';
			loadJsonMetadata(path, function(err, data) {
				if(err) {
					res.status(404).json({
						error: util.format(
							"ID does not exist: %s", req.params.id)
					});
					return;
				}

				// If the request is an external CORS request we'll restrict
				// access to items which are not embeddable. This prevents
				// external sites using CORS request to get at non-embeddable
				// content, while allowing cudl itself to get at it. Note that
				// there's nothing to stop someone setting up a proxy which
				// strips or fakes the origin header.
				if(serviceUtil.isExternalCorsRequest(req) &&
						data.embeddable === false) {
					res.status(403).json({
						error: "This item is only available from " +
								"cudl.lib.cam.ac.uk"
					});
					return;
				}

				res.json(data);
			});
		} else {
			path = config.dataDir+'/data/'+req.params.format+'/'+req.params.id+'/'+req.params.id+'.xml';
 			res.sendfile(path);
  		}
	}
);

function loadJsonMetadata(path, cb) {
	fs.readFile(path, "utf-8", function(err, data) {
		if(err) cb(err);
		try{
			var parsed = JSON.parse(data);
		}
		catch(e) {
			cb(util.format("Error parsing file as JSON: %s. %s", path, e));
			return;
		}
		cb(null, parsed);
	});
}

module.exports = router;
