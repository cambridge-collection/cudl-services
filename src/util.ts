import escapeStringRegexp from 'escape-string-regexp';
import { Request } from 'express';
import url from 'url';
import * as util from 'util';
import { ValueError } from './errors';

const CUDL_HOST = 'cudl.lib.cam.ac.uk';
const CUDL_HOST_REGEX = new RegExp(
  '(?:^|\\.)' + escapeStringRegexp(CUDL_HOST) + '$'
);

export const CORS_HEADERS = Object.freeze({
  'Access-Control-Allow-Origin': '*',
});

export function isExternalCorsRequest(req: Request) {
  const origin = req.header('origin');
  if (!origin) {
    return false;
  }

  const host = url.parse(origin).hostname;

  // If we have an origin header and it's not cudl, then it's an external cors
  // request.
  return typeof host === 'string' && !CUDL_HOST_REGEX.test(host);
}

export function isSimplePathSegment(value: string): boolean {
  return /^[\w -]+$/.test(value);
}

/**
 * A type guard which narrows strings to members of string-valued enums.
 *
 * @param _enum The enum type
 * @param val The value to narrow
 */
export function isEnumMember<E extends string>(
  _enum: { [k in string]: E },
  val: string
): val is E {
  const members = Object.values(_enum) as unknown[];
  return members.includes(val);
}

export function requireRequestParam(req: Request, param: string): string {
  const value = req.params[param] as typeof req.params[string] | undefined;
  if (typeof value !== 'string') {
    throw new Error(`Request has no value for param ${util.inspect(param)}`);
  }
  return value;
}

export function requireRequestParams<T extends string>(
  req: Request,
  ...params: T[]
): { [key in T]: string } {
  const obj = {} as { [key in T]: string };
  for (const param of params) {
    obj[param] = requireRequestParam(req, param);
  }
  return obj;
}

export function factory<A, B>(
  constructor: new (...args: A[]) => B
): (...args: A[]) => B {
  return (...args: A[]) => new constructor(...args);
}

/** The type of the only argument of a 1-ary constructor. */
export type UnaryConstructorArg<
  T extends new (arg: unknown) => unknown
> = ConstructorParameters<T>[0];

export type ComparePrimitive = string | number | boolean | ComparePrimitive[];
export type CompareModifier = [Direction, CompareValue];
export type CompareValue = ComparePrimitive | CompareModifier | CompareValue[];

export function compare(a: CompareValue, b: CompareValue): number {
  if (typeof a !== typeof b) {
    throw new Error(
      `compare() not supported between ${typeof a} and ${typeof b}`
    );
  }

  if (!Array.isArray(a)) {
    return a === b ? 0 : a < b ? -1 : 1;
  }

  // Directional compare modifier
  if ((a.length > 0 && a[0] === asc) || a[0] === desc) {
    b = b as CompareModifier | CompareValue[];
    if (a.length !== 2) {
      throw new Error(`Invalid compare direction modifier: modifiers must be a pair of [direction, value], \
got: ${util.inspect(a)}`);
    }
    if (b.length !== 2 || b[0] !== a[0]) {
      throw new Error(
        `Invalid compare direction modifier: modifiers on both comparison values must use the same direction`
      );
    }

    return compare(a[1], b[1]) * (a[0] === desc ? -1 : 1);
  }

  a = a as CompareValue[];
  b = b as CompareValue[];

  for (let i = 0; i < Math.min(a.length, b.length); ++i) {
    const order = compare(a[i], b[i]);
    if (order !== 0) {
      return order;
    }
  }
  // shorter array is less
  return a.length - b.length;
}

const asc: unique symbol = Symbol('compare.asc');
const desc: unique symbol = Symbol('compare.desc');
compare.asc = asc;
compare.desc = desc;
type Direction = typeof asc | typeof desc;

export function sorted<T extends ComparePrimitive>(items: Iterable<T>): T[];
export function sorted<T>(
  items: Iterable<T>,
  key: (value: T) => CompareValue
): T[];
export function sorted<T>(
  items: Iterable<T>,
  key?: (value: T) => CompareValue
): T[] {
  if (key === undefined) {
    const sorted = Array.from(
      (items as Iterable<unknown>) as Iterable<ComparePrimitive>
    );
    sorted.sort((a, b) => compare(a, b));
    return (sorted as unknown[]) as T[];
  }

  const indexedItems = index(items);
  const keys = indexedItems.map(([_, value]) => key(value));
  indexedItems.sort(([ia], [ib]) => compare(keys[ia], keys[ib]));

  // Remove indexes in place
  const sorted = (indexedItems as unknown) as T[];
  for (let i = 0; i < indexedItems.length; ++i) {
    sorted[i] = indexedItems[i][1];
  }

  return sorted;
}

function index<T>(items: Iterable<T>): Array<[number, T]> {
  const index: Array<[number, T]> = [];
  let i = 0;
  for (const item of items) {
    index[i] = [i, item];
    i++;
  }
  return index;
}
