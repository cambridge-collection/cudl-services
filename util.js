var url = require("url");
var regexpQuote = require("regexp-quote");


var CUDL_HOST = "cudl.lib.cam.ac.uk";
var CUDL_HOST_REGEX = new RegExp("(?:^|\\.)" + regexpQuote(CUDL_HOST) + "$");

var CORS_HEADERS = module.exports.CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*"
};


module.exports.isExternalCorsRequest = function isExternalCorsRequest(req) {
    var origin = req.header("origin") || "";
    if(!origin) { return false; }

    var host = url.parse(origin).hostname;

    // If we have an origin header and it's not cudl, then it's an external cors
    // request.
    return !CUDL_HOST_REGEX.test(host);
};
