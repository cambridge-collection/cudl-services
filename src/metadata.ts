import * as assert from 'assert';
import { AssertionError } from 'assert';
import fs from 'fs';
import path from 'path';
import * as util from 'util';
import { promisify } from 'util';
import {
  BaseError,
  isEnumMember,
  isSimplePathSegment,
  NotFoundError,
} from './util';

export class MetadataError extends BaseError {}

export interface MetadataRepository<F extends string = string> {
  getPath(format: F, id: string): Promise<string>;
  getBytes(format: F, id: string): Promise<Buffer>;
}

export enum CUDLFormat {
  DCP = 'dcp',
  EAD = 'ead',
  ESSAY = 'essay',
  MODS = 'mods',
  TEI = 'tei',
  TRANSCRIPTION = 'transcription',
  JSON = 'json',
}

abstract class BaseMetadataRepository<T extends string>
  implements MetadataRepository<T> {
  abstract getPath(format: T, id: string): Promise<string>;

  async getBytes(format: T, id: string): Promise<Buffer> {
    const path = await this.getPath(format, id);
    try {
      return await promisify(fs.readFile)(path);
    } catch (e) {
      throw new MetadataError(
        `Failed to load metadata from ${path}: ${e.message}`,
        e
      );
    }
  }
}

export interface CUDLMetadataRepository extends MetadataRepository<CUDLFormat> {
  getJSON(id: string): Promise<ItemJSON>;
}

export class DefaultCUDLMetadataRepository
  extends BaseMetadataRepository<CUDLFormat>
  implements CUDLMetadataRepository {
  private readonly dataDir: string;

  constructor(dataDir: string) {
    super();
    this.dataDir = dataDir;
  }

  async getPath(format: CUDLFormat, id: string) {
    assert.ok(isSimplePathSegment(format));

    if (format === CUDLFormat.TRANSCRIPTION) {
      const idParts = /^([\w-]+)\/([\w-]+)(?:\.xml)?$/.exec(id);
      if (!idParts) {
        throw new Error(`Invalid ${CUDLFormat.TRANSCRIPTION} id: ${id}`);
      }
      return path.join(
        this.dataDir,
        'data',
        format,
        idParts[1],
        `${idParts[2]}.xml`
      );
    }

    if (!isSimplePathSegment(id)) {
      throw new Error(
        `${name} is not a valid path segment: ${util.inspect(id)}`
      );
    }

    if (format === CUDLFormat.JSON) {
      return path.join(this.dataDir, 'json', `${id}.json`);
    }
    return path.join(this.dataDir, 'data', format, id, `${id}.xml`);
  }

  async getJSON(id: string): Promise<ItemJSON> {
    const jsonPath = await this.getPath(CUDLFormat.JSON, id);
    try {
      const content = await promisify(fs.readFile)(jsonPath, 'utf-8');
      const data = JSON.parse(content);
      if (!isItemJSON(data)) {
        throw new MetadataError('unexpected JSON structure');
      }
      return data;
    } catch (e) {
      throw new MetadataError(
        `Failed to load metadata from ${jsonPath}: ${e.message}`,
        e
      );
    }
  }
}

export interface ItemJSON {
  embeddable?: boolean;
  descriptiveMetadata?: [
    {
      metadataRights?: string;
    }
  ];
}

export function isItemJSON(data: any): data is ItemJSON {
  if (typeof data !== 'object') {
    return false;
  }

  return (
    (data.embeddable === undefined || typeof data.embeddable === 'boolean') &&
    Array.isArray(data.descriptiveMetadata) &&
    (data.descriptiveMetadata || []).every(
      (dmd: any) =>
        (typeof dmd === 'object' && dmd.metadataRights === undefined) ||
        typeof dmd.metadataRights === 'string'
    )
  );
}

export enum LegacyDarwinFormat {
  DEFAULT = 'dcpfull',
}

export class LegacyDarwinMetadataRepository extends BaseMetadataRepository<
  LegacyDarwinFormat.DEFAULT
> {
  private readonly pathResolver: PathResolver;

  constructor(dataDir: string) {
    super();
    this.pathResolver = createLegacyDarwinPathResolver(dataDir);
  }

  async getPath(id: string): Promise<string>;
  async getPath(
    format: LegacyDarwinFormat.DEFAULT,
    id: string
  ): Promise<string>;
  async getPath(format: string, id?: string) {
    if (id === undefined) {
      id = format;
      format = LegacyDarwinFormat.DEFAULT;
    }
    assert.ok(format === LegacyDarwinFormat.DEFAULT);
    return this.pathResolver(id);
  }
}

type PathResolver = (id: string) => Promise<string>;

interface CachedIDMap {
  seq: number;
  paths: Map<string, string>;
  dirModifiedTime: number;
  expirationTime: number;
}

export function createLegacyDarwinPathResolver(dataDir: string): PathResolver {
  const cacheTTL = 60 * 1000;
  let seq = 0;
  let cachedIDMap: CachedIDMap | undefined;
  let nextCachedIDMap: Promise<CachedIDMap> | undefined;

  const getCachedIDMap = async (now: number) => {
    const dirModifiedTime = (await promisify(fs.stat)(dataDir)).mtimeMs;
    return {
      seq: seq++,
      dirModifiedTime,
      expirationTime: now + cacheTTL,
      paths: await getLegacyDarwinMetadataPaths(dataDir),
    };
  };

  return async (id): Promise<string> => {
    const now = Date.now();

    if (cachedIDMap === undefined && nextCachedIDMap === undefined) {
      nextCachedIDMap = getCachedIDMap(now);
    }

    const _cachedIDMap = await Promise.race(
      [nextCachedIDMap, cachedIDMap].filter(x => x !== undefined)
    );
    if (_cachedIDMap === undefined) {
      throw new AssertionError();
    }
    if (cachedIDMap === undefined || cachedIDMap.seq !== _cachedIDMap.seq) {
      cachedIDMap = _cachedIDMap;
      nextCachedIDMap = undefined;
    }

    if (now > _cachedIDMap.expirationTime && nextCachedIDMap === undefined) {
      nextCachedIDMap = getCachedIDMap(now);
    }

    const relPath = _cachedIDMap.paths.get(id);
    if (relPath === undefined) {
      throw new NotFoundError(`no metadata found for id: ${id}`);
    }
    return path.resolve(dataDir, relPath);
  };
}

async function getLegacyDarwinMetadataPaths(
  dataDir: string
): Promise<Map<string, string>> {
  const files = await promisify(fs.readdir)(dataDir);
  files.sort();

  const paths = new Map();
  for (const f of files) {
    const match = /^([a-zA-Z0-9]+)_.*\.xml$/.exec(f);

    if (match) {
      const id = match[1];
      if (!paths.has(id)) {
        paths.set(id, f);
      }
    }
  }

  return paths;
}
