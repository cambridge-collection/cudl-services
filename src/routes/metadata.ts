import express, {Response} from 'express';
import expressAsyncHandler from 'express-async-handler';
import {StatusCodes} from 'http-status-codes';
import util from 'util';
import {
  IsExternalAccessPermitted,
  IsExternalEmbedPermitted,
  ItemJSON,
  MetadataPredicate,
  MetadataProvider,
  MetadataResponse,
} from '../metadata';

import {
  applyLazyDefaults,
  CORS_HEADERS,
  ExternalCorsRequestMatcher,
  isEnumMember,
  isExternalCorsRequest,
  isSimplePathSegment,
} from '../util';
import {CUDLFormat, CUDLMetadataRepository} from '../metadata/cudl';
import {ErrorCategories, isTagged} from '../errors';

export interface MetadataResponseEmitter {
  canEmit(metadataResponse: MetadataResponse): boolean;
  emit(metadataResponse: MetadataResponse, resp: Response): Promise<void>;
}

export interface GetRoutesV2Options {
  router?: express.Router;
  metadataProviders: ReadonlyMap<string, MetadataProvider>;
  isExternalEmbedPermitted?: MetadataPredicate;
  isExternalAccessPermitted?: MetadataPredicate;
  isExternalCorsRequest: ExternalCorsRequestMatcher;
  metadataEmitters?: Array<MetadataResponseEmitter>;
}

type CreateMetadataHandlerV2Options = Omit<
  Required<GetRoutesV2Options>,
  'router'
>;

export function getRoutesV2(options: GetRoutesV2Options) {
  const {router, ...handlerOptions} = applyLazyDefaults(options, {
    router: () => express.Router(),
    isExternalEmbedPermitted: () => IsExternalEmbedPermitted,
    isExternalAccessPermitted: () => IsExternalAccessPermitted,
    metadataEmitters: () => [],
  });

  /* GET home page. */
  router.get('/', (req, res) => {
    res.status(StatusCodes.UNAUTHORIZED).send('Unathorised');
  });

  router.get('/:format/:id', createMetadataHandlerV2(handlerOptions));

  return router;
}

function createMetadataHandlerV2(options: CreateMetadataHandlerV2Options) {
  return expressAsyncHandler(async (req, res) => {
    // We always want to allow remote ajax access
    res.set(CORS_HEADERS);

    // The response depends on the Origin header, as we block access to
    // non-embeddable items from non-cudl origins. If we don't set
    // Vary: Origin then a response for a client on CUDL could be used by a
    // cache to service a request from an external site.
    res.set('Vary', 'Origin');

    // Ensure our path vars are safe to build FS paths from
    const id = req.params.id;
    const format = req.params.format;
    if (!isSimplePathSegment(id)) {
      res.status(StatusCodes.BAD_REQUEST).json({
        error: util.format(`Bad id: ${id}`),
      });
      return;
    }

    const provider = options.metadataProviders.get(format);
    if (!provider) {
      res.status(StatusCodes.BAD_REQUEST).json({
        error: `Bad format: ${format}`,
      });
      return;
    }

    let metadata: MetadataResponse;
    try {
      metadata = await provider.query(id);
    } catch (e) {
      if (isTagged(e) && new Set(e.tags).has(ErrorCategories.NotFound)) {
        res.status(StatusCodes.NOT_FOUND).json({
          error: `ID does not exist: ${id}`,
        });
        return;
      }
      throw e;
    }

    // If the request is an external CORS request we'll restrict
    // access to items which are not embeddable. This prevents
    // external sites using CORS request to get at non-embeddable
    // content, while allowing cudl itself to get at it. Note that
    // there's nothing to stop someone setting up a proxy which
    // strips or fakes the origin header.
    if (
      options.isExternalCorsRequest(req) &&
      (await options.isExternalEmbedPermitted(metadata)) === false
    ) {
      res.status(StatusCodes.FORBIDDEN).json({
        error: `This metadata is only available from ${options.isExternalCorsRequest.internalDomainNameMatcher.describeMatchingDomains()}`,
      });
      return;
    }

    if ((await options.isExternalAccessPermitted(metadata)) === false) {
      res.status(StatusCodes.FORBIDDEN).json({
        error: util.format('Access not allowed to requested metadata'),
      });
      return;
    }

    for (const emitter of options.metadataEmitters) {
      if (emitter.canEmit(metadata)) {
        await emitter.emit(metadata, res);
        return;
      }
    }
    // Send metadata as UTF-8 text if not specifically handled
    res.contentType('text/plain; charset=utf-8');
    res.send((await metadata.getBytes()).toString('utf-8'));
  });
}

export function getRoutes(options: {
  router?: express.Router;
  metadataRepository: CUDLMetadataRepository;
}) {
  const router = options.router || express.Router();

  /* GET home page. */
  router.get('/', (req, res) => {
    res.status(StatusCodes.UNAUTHORIZED).send('Unathorised');
  });

  router.get('/:format/:id', createMetadataHandler(options.metadataRepository));

  return router;
}

function createMetadataHandler(metadataRepository: CUDLMetadataRepository) {
  return expressAsyncHandler(
    async (req: express.Request, res: express.Response) => {
      // We always want to allow remote ajax access
      res.set(CORS_HEADERS);

      // The response depends on the Origin header, as we block access to
      // non-embeddable items from non-cudl origins. If we don't set
      // Vary: Origin then a response for a client on CUDL could be used by a
      // cache to service a request from an external site.
      res.set('Vary', 'Origin');

      // Ensure our path vars are safe to build FS paths from
      const id = req.params.id;
      const format = req.params.format;
      if (!isSimplePathSegment(id)) {
        res.status(StatusCodes.BAD_REQUEST).json({
          error: util.format(`Bad id: ${id}`),
        });
        return;
      }

      if (!isEnumMember(CUDLFormat, format)) {
        res.status(StatusCodes.BAD_REQUEST).json({
          error: util.format(`Bad format: ${format}`),
        });
        return;
      }

      let item: ItemJSON;
      try {
        item = await metadataRepository.getJSON(id);
      } catch (e) {
        if (e?.nested?.code === 'ENOENT') {
          res.status(StatusCodes.NOT_FOUND).json({
            error: `ID does not exist: ${id}`,
          });
          return;
        } else {
          throw e;
        }
      }

      if (req.params.format === 'json') {
        // If the request is an external CORS request we'll restrict
        // access to items which are not embeddable. This prevents
        // external sites using CORS request to get at non-embeddable
        // content, while allowing cudl itself to get at it. Note that
        // there's nothing to stop someone setting up a proxy which
        // strips or fakes the origin header.
        if (
          isExternalCorsRequest(req) &&
          typeof item.embeddable === 'boolean' &&
          !item.embeddable
        ) {
          res.status(StatusCodes.FORBIDDEN).json({
            error: 'This item is only available from ' + 'cudl.lib.cam.ac.uk',
          });
          return;
        }

        res.json(item);
      } else {
        // This returns the original metadata.  We only want to return the metadata if
        // the metadataRights field is present (and non-empty) in the JSON.
        if (item?.descriptiveMetadata?.[0]?.metadataRights?.trim()) {
          // Return metadata
          res.contentType('text/plain');
          res.sendFile(await metadataRepository.getPath(format, id));
        } else {
          res.status(StatusCodes.FORBIDDEN).json({
            error: util.format('Access not allowed to requested metadata'),
          });
        }
      }
    }
  );
}
