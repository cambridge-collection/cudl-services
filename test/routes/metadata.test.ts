import express from 'express';
import fs from 'fs';
import {StatusCodes} from 'http-status-codes';
import path from 'path';
import request from 'supertest';
import {promisify} from 'util';
import {CUDLMetadataRepository} from '../../src/metadata';

import {getRoutes} from '../../src/routes/metadata';
import {TEST_DATA_PATH} from '../constants';

import {getTestDataMetadataRepository} from '../utils';

function getTestApp(metadataRepository: CUDLMetadataRepository) {
  const app = express();
  app.use('/', getRoutes({metadataRepository}));
  return app;
}

describe('metadata routes /:format/:id', () => {
  let repo: CUDLMetadataRepository, app: express.Application;

  beforeEach(() => {
    repo = getTestDataMetadataRepository();
    app = getTestApp(repo);
  });

  test.each([
    ['MS-ADD-03959'],
    ['non-embeddable-item'],
    ['item-without-rights-statement'],
  ])('/json/%s reponds with JSON metadata', async id => {
    const response = await request(app).get(`/json/${id}`);
    expect(response.status).toBe(StatusCodes.OK);
    expect(response.get('content-type')).toBe(
      'application/json; charset=utf-8'
    );
    expect(response.body).toEqual(
      JSON.parse(
        await promisify(fs.readFile)(
          path.resolve(TEST_DATA_PATH, `metadata/json/${id}.json`),
          'utf-8'
        )
      )
    );
  });

  test('/json/:id forbids cross-origin requests for non-embeddable items', async () => {
    const response = await request(app)
      .get('/json/non-embeddable-item')
      .set('Origin', 'https://example.com');
    expect(response.status).toBe(StatusCodes.FORBIDDEN);
    expect(response.body.error).toMatch(
      'This item is only available from cudl.lib.cam.ac.uk'
    );
  });

  test('/json/:id responds with 404 for missing ID', async () => {
    const response = await request(app).get('/json/MS-MISSING');
    expect(response.status).toBe(StatusCodes.NOT_FOUND);
  });

  test('/json/:id responds with 500 for invalid metadata', async () => {
    const response = await request(app).get('/json/INVALID');
    expect(response.status).toBe(StatusCodes.INTERNAL_SERVER_ERROR);
  });

  test.each([
    ['format', `/${encodeURIComponent('abc/def')}/foo`],
    ['id', `/foo/${encodeURIComponent('abc/def')}`],
  ])('%s param cannot contain slashes', async (param, url) => {
    const response = await request(app).get(url);
    expect(response.status).toBe(StatusCodes.BAD_REQUEST);
  });

  test('/:format/:id responds with 500 for invalid metadata when requesting non-json metadata', async () => {
    const response = await request(app).get('/tei/INVALID');
    expect(response.status).toBe(StatusCodes.INTERNAL_SERVER_ERROR);
  });

  test('/:format/:id responds with non-JSON metadata', async () => {
    const response = await request(app).get('/tei/MS-ADD-03959');
    expect(response.status).toBe(StatusCodes.OK);
    expect(response.get('content-type')).toBe('text/plain; charset=utf-8');
    expect(response.text).toEqual(
      await promisify(fs.readFile)(
        path.resolve(
          TEST_DATA_PATH,
          'metadata/data/tei/MS-ADD-03959/MS-ADD-03959.xml'
        ),
        'utf-8'
      )
    );
  });

  test('/:format/:id responds with 403 for non-json metadata without a rights statement', async () => {
    const response = await request(app).get(
      '/tei/item-without-rights-statement'
    );
    expect(response.status).toBe(StatusCodes.FORBIDDEN);
    expect(response.body.error).toBe(
      'Access not allowed to requested metadata'
    );
  });
});
