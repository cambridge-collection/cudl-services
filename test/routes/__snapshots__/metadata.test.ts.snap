// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata routes v2 /:format/:id /:format/:id responds with NOT_FOUND when MetadataProvider throws an error tagged with NotFound 1`] = `
Object {
  "error": "ID does not exist: 123",
}
`;
